File: 3759-choose-k-elements-with-maximum-sum/3759-choose-k-elements-with-maximum-sum.cpp
class Solution {
public:
    vector<long long> findMaxSum(vector<int>& nums1, vector<int>& nums2,
                                 int k) {
        map<int, vector<int>> mp;
        for (int i = 0; i < nums1.size(); ++i) {
            mp[nums1[i]].push_back(i);
        }
        priority_queue<int, vector<int>, greater<int>> pq;
        long long sum = 0;
        vector<long long> ans(nums1.size());
        for (auto& i : mp) {
            int temp = sum;
            for (auto& j : i.second) {
                ans[j] = sum;
            }
            for (auto& j : i.second) {
                sum += nums2[j];
                pq.push(nums2[j]);
                if (pq.size() > k) {
                    sum -= pq.top();
                    pq.pop();
                }
            }
        }
        return ans;
    }
};

File: 3759-choose-k-elements-with-maximum-sum/3759-choose-k-elements-with-maximum-sum.java
class Solution {
    public long[] findMaxSum(int[] nums1, int[] nums2, int k) {
        int n = nums1.length;
        long[] ans = new long[n];

        TreeMap<Integer, List<Integer>> indexMap = new TreeMap<>();
        for (int i = 0; i < n; i++) {
            indexMap.computeIfAbsent(nums1[i], v -> new ArrayList<>()).add(i);
        }

        PriorityQueue<Integer> pq = new PriorityQueue<>();
        long sum = 0;

        for (Map.Entry<Integer, List<Integer>> entry : indexMap.entrySet()) {
            List<Integer> indices = entry.getValue();

            for (int idx : indices) {
                ans[idx] = sum;
            }

            for (int idx : indices) {
                pq.add(nums2[idx]);
                sum += nums2[idx];

                if (pq.size() > k) {
                    sum -= pq.poll();
                }
            }
        }

        return ans;
    }
}